network/openssl_privatekey.py

```diff
@@ -0,0 +1,204 @@
+- openssl_privatekey: name=ansible.com.pem path=/etc/ssl/private type=DSA
+'''
+
+RETURN = '''
+size:
+    description: Size of the SSL private key
+    returned:
+        - changed
+        - success
+    type: integer
+    sample: 4096
+type:
+    description: Algorithm used to generate the SSL private key
+    returned:
+        - changed
+        - success
+    type: string
+    sample: RSA
+privatekey:
+    description: Path to the generated SSL private key
+    returned:
+        - changed
+        - success
+    type: string
+    sample: /etc/ssl/private/ansible.com.pem
+'''
+
+class PrivateKeyError(Exception):
+    pass
+
+
+class PrivateKey(object):
+
+    def __init__(self, module):
+        self.size = module.params['size']
+        self.state = module.params['state']
+        self.name = module.params['name']
+        self.type = module.params['type']
+        self.force = module.params['force']
+        self.path = module.params['path']
+        self.changed = True
+
+    @property
+    def file_path(self):
+        return os.path.join(self.path, self.name)
+   
+    def generate(self):
+        '''Generate a keypair.'''
+
+        if not self.force and os.path.exists(self.file_path):
+            self.changed = False
+            return
+
+        self.privatekey = crypto.PKey()
+
+        if self.type == 'RSA':
+            crypto_type = crypto.TYPE_RSA
+        else:
+            crypto_type = crypto.TYPE_DSA
+
+        try:
+            self.privatekey.generate_key(crypto_type, self.size)
+            privatekey_file = open(self.file_path, 'w')
+            privatekey_file.write(crypto.dump_privatekey(crypto.FILETYPE_PEM, self.privatekey))
+            privatekey_file.close()
+        except TypeError:
+            raise PrivateKeyError()
+        except ValueError:
+            raise PrivateKeyError()
+
+    def remove(self):
+        '''Remove the private key from the filesystem.'''
+
+        try:
+            os.remove(self.file_path)
+        except OSError:
+            self.changed = False
+
+    def dump(self):
+        '''Serialize the object into a dictionnary.'''
+
+        result = {
+            'size': self.size,
+            'type': self.type,
+            'privatekey': self.file_path,
+            'changed': self.changed,
+        }
+
+        return result
+        
+
+def main():
+
+    module = AnsibleModule(
+        argument_spec = dict(
+            state=dict(default='present', choices=['present', 'absent'], type='str'),
+            name=dict(required=True, type='str'),
+            size=dict(default=4096, type='int'),
+            type=dict(default='RSA', choices=['RSA', 'DSA'], type='str'),
+            force=dict(default=False, type='bool'),
+            path=dict(required=True, type='str'),
+        )
+    )
+
+    path=module.params['path']
+    if not os.path.isdir(path):
+        module.fail_json(name=path, msg='The directory %s does not exist' % path)
```

Comment:
Error message is not exactly right, since this could be a file, so what about adding "or is not a directory" ?